### PR TITLE
batch-model-training

### PR DESCRIPTION
**Pull request by alphasentaurii**
semi-automated data ingest and compute resource prediction model training (memory bin classifier, memory regressor and wallclock regressor) via AWS batch. Data ingest and model re-training for large HST reprocessing runs can be deployed in one step by submitting a job in Batch. Docker image for batch compute env, job definition, and job queue are deployed via terraform.

### CODE REVIEW

File: modeling/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM library/ubuntu:20.04
+#SSL/TLS cert setup for STScI AWS firewalling
+USER root
+ENV REQUESTS_CA_BUNDLE=/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem
+ENV CURL_CA_BUNDLE=/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem
+RUN mkdir -p /etc/ssl/certs/ && mkdir -p /etc/pki/ca-trust/extracted/pem/
+COPY certs/tls-ca-bundle.pem /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem
+RUN ln -s /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem /etc/ssl/certs/ca-bundle.crt && ln -s /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem /etc/ssl/certs/ca-certificates.crt && mkdir -p /etc/pki/ca-trust/extracted/openssl
+RUN apt update && apt install -y python3-dev python3-pip curl unzip && update-alternatives --install /usr/bin/python python /usr/bin/python3 10 && curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip" && unzip awscliv2.zip && ./aws/install
```
Comment (bhayden53) [inline modeling/Dockerfile:9]: I suspect this is why ECR is able to complain about python in your image, whereas the CALDP image has python in conda.

Not suggesting any changes right now, just thinking out loud on the ECR scan results.

File: terraform/model_training.tf [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  compute_resources {
+    allocation_strategy = "BEST_FIT"
+    instance_role = data.aws_ssm_parameter.ecs_instance_role.value
+    type = "EC2"
+    bid_percentage = 0
+    tags = {}
+    subnets             = local.batch_subnet_ids
+    security_group_ids  = local.batch_sgs
+    instance_type = ["c5a.2xlarge"]
```
Comment (bhayden53) [inline terraform/model_training.tf:33]: curious what went into this decision. Do you use all the cores? Or do you need the memory?

**Review by jaytmiller — APPROVED** [review]
First time looking at this stuff but LGTM.

File: modeling/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+tensorflow==2.5.0
+sklearn==0.0
```
Comment (bhayden53) [inline modeling/requirements.txt:2]: this seems odd. I googled it and came across this:
https://stackoverflow.com/a/40167445/1502840

Should there be an actual version number there and ubuntu gave something faulty at some point?

Comment (bhayden53) [reply modeling/requirements.txt:2]: ah, I guess that really is the version of sklearn. Silly. 

I was just trying to understand if setting the version to 0.0 was some kind of special flag to pip.

File: terraform/model_training.tf [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    min_vcpus = 0
+    desired_vcpus = 0
+  }
+  lifecycle {
+    ignore_changes = [compute_resources.0.desired_vcpus]
+    create_before_destroy = true
+  }
+}
+
+resource "aws_batch_job_definition" "model_job_def_main" {
```
Comment (bhayden53) [inline terraform/model_training.tf:44]: how does the job get triggered? Should there be a cron-like cloudwatch event (i.e. like the blackboard lambda)?

Comment (alphasentaurii) [reply terraform/model_training.tf:44]: yes - we can add this later as a separate feature but for now we would manually submit as a batch job when needed

**Pull request merged**